File: CSC205-Cpp/Module10-Searching/debug01/binary_search.cpp
// binary_search.cpp
// Binary search over a sorted array of part numbers. For each requested part
// the program reports where it was found and how many elements it examined.

#include <iostream>
using namespace std;

// Returns the index of target in the sorted array a[0..n-1], or -1 if it is
// not present. comparisons is set to the number of elements examined.
int binarySearch(const int a[], int n, int target, int& comparisons) {
    int low = 0;
    int high = n - 1;
    comparisons = 0;

    while (low <= high) {
        int mid = low + (high - low) / 2;
        comparisons++;
        if (a[mid] == target) {
            return mid;
        } else if (a[mid] < target) {
            low = mid + 1;
        } else {
            high = mid;
        }
    }
    return -1;
}

int main() {
    const int n = 10;
    int parts[n] = { 104, 211, 305, 418, 522, 637, 741, 856, 960, 1073 };
    int requests[] = { 522, 104, 1073, 741, 300, 2000 };

    for (int target : requests) {
        int comparisons = 0;
        int index = binarySearch(parts, n, target, comparisons);
        if (index >= 0) {
            cout << "part " << target << ": found at index " << index
                 << " (" << comparisons << " comparisons)" << endl;
        } else {
            cout << "part " << target << ": not found"
                 << " (" << comparisons << " comparisons)" << endl;
        }
    }
    return 0;
}

File: CSC205-Java/Module10-Searching/debug01/BinarySearch.java
// Binary search over a sorted array of exam scores.
// Reports the index where each target lives, or -1 if it is absent.
public class BinarySearch {

    // Returns the index of target in the sorted array a, or -1 if not present.
    public static int binarySearch(int[] a, int target) {
        int low = 0;
        int high = a.length - 1;

        while (low <= high) {
            int mid = (low + high) / 2;
            if (a[mid] == target) {
                return mid;
            } else if (a[mid] < target) {
                low = mid + 1;
            } else {
                high = mid;
            }
        }
        return -1;
    }

    // Plain linear search, used to double-check the results.
    public static int linearSearch(int[] a, int target) {
        for (int i = 0; i < a.length; i++) {
            if (a[i] == target) {
                return i;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        int[] scores = {3, 8, 15, 21, 27, 34, 42, 50};
        int[] targets = {21, 3, 50, 40};

        System.out.print("Scores:");
        for (int s : scores) {
            System.out.print(" " + s);
        }
        System.out.println();

        for (int t : targets) {
            int index = binarySearch(scores, t);
            int check = linearSearch(scores, t);
            String where = (index < 0) ? "not found" : "index " + index;
            System.out.println("search " + t + ": " + where
                    + "   (linear search says " + check + ")");
        }
    }
}
